pyscf/lib/dft/libxc_itrf.c

```diff
@@ -362,6 +363,124 @@ static void _eval_xc(xc_func_type *func_x, int spin, int np,
+                if (xc_func_init(&func, id, XC_UNPOLARIZED) != 0){
+                        fprintf(stderr, "XC functional %d not found\n", id);
+                        raise_error -1;
+                }
+                int ref_size = func.info->ext_params.n;
+                xc_func_end(&func);
+                if (ref_size != size) {
+                        fprintf(stderr, "Inconsistent parameter size. Expected %d but got %lu.\n", ref_size, size);
+                        raise_error -1;
+                }
+        }
+
+        // Create new entry
+        int i = custom_ext_params_count++;
+        custom_ext_params = realloc(custom_ext_params, custom_ext_params_count * sizeof(ext_params));
+        ext_params *p = &custom_ext_params[i];
+        p->id = id;
+        p->params = malloc(size * sizeof(double));
+        memcpy(p->params, params, size * sizeof(double));
+        p->size = size;
+        return 0;
+}
+
+int remove_ext_params(int id)
+{
+        for (int i = 0; i < custom_ext_params_count; i++) {
+                if (custom_ext_params[i].id == id) {
+                        free(custom_ext_params[i].params);
+                        for (int j = i; j < custom_ext_params_count - 1; j++) {
+                                custom_ext_params[j] = custom_ext_params[j + 1];
+                        }
+                        custom_ext_params_count--;
+                        return 0;
+                }
+        }
+        return -1;
+}
+
+void clear_ext_params()
+{
+        for (int i = 0; i < custom_ext_params_count; i++) {
+                free(custom_ext_params[i].params);
+        }
+        free(custom_ext_params);
+        custom_ext_params = NULL;
+        custom_ext_params_count = 0;
+}
+
+void print_ext_params()
+{
+        puts("Custom external functionals defined for the following functional(s):");
+        for (int i = 0; i < custom_ext_params_count; i++) {
+                ext_params *p = &custom_ext_params[i];
+                xc_func_type func;
+                if (xc_func_init(&func, p->id, XC_UNPOLARIZED) != 0){
+                        fprintf(stderr, "XC functional %d not found\n", p->id);
+                        raise_error;
+                }
+                printf("name: %s\n", func.info->name);
+                printf("id: %d\n", p->id);
+                printf("parameter size: %lu\n", p->size);
+                puts("parameter values:");
+                printf("%3s %14s %8s %s\n", "idx", "value", "name", "description");
+                for (int j = 0; j < p->size; j++) {
+                        printf("%3i %14.6e %8s %s\n", j, p->params[j],
+                                xc_func_info_get_ext_params_name(func.info, j),
+                                xc_func_info_get_ext_params_description(func.info, j));
+                }
+                xc_func_end(&func);
+                putchar('\n');
+        }
+}
+
+static int _xc_func_init(xc_func_type *func, int xc_id, int spin)
```

Comment:
This is also named too closely to `xc_func_init`, it should be clearer that this is a function in the PySCF wrapper. Use e.g. `xc_init_wrapper()`

Reply:
I renamed it as `init_xc_and_set_ext_params()`. This should be clear and distinguishable from Libxc functions.